dbt/adapters/duckdb/plugins/native.py

```diff
@@ -0,0 +1,108 @@
+import os
+from typing import Any
+from typing import Dict
+
+from duckdb import DuckDBPyRelation
+
+from . import BasePlugin
+from ..utils import SourceConfig
+from ..utils import TargetConfig
+
+# here will be parquet,csv,json implementation,
+# this plugin should be default one if none is specified
+# we can change the name of the plugin
+
+
+class Plugin(BasePlugin):
+    def initialize(self, config: Dict[str, Any]):
+        pass
+
+    def configure_cursor(self, cursor):
+        pass
+
+    def default_materialization(self):
+        return "view"
+
+    # this one can be better not to go over some other format and df but directly
+    # https://stackoverflow.com/questions/78055585/how-to-reference-duckdbpyrelation-from-another-connection
+    def load(self, source_config: SourceConfig, coursor=None):
+        location = external_read_location(
+            source_config.meta.get("location", "").get("path"),
+            source_config.meta.get("config", {}).get("options", {}),
+        )
+        return f"(SELECT * FROM '{location}')"
+
+    def can_be_upstream_referenced(self):
+        return True
+
+    def create_source_config(self, target_config: TargetConfig) -> SourceConfig:
+        source_config = SourceConfig(
+            name=target_config.relation.name,
+            identifier=target_config.relation.identifier,
+            schema=target_config.relation.schema,
+            database=target_config.relation.database,
+            meta=target_config.as_dict(),
+            tags=[],
+        )
+        return source_config
+
+    def store(self, df: DuckDBPyRelation, target_config: TargetConfig, cursor=None):
+        location = target_config.location.path
+        options = external_write_options(location, target_config.config.get("options", {}))
+        cursor.sql(f"COPY (SELECT * FROM df) to '{location}' ({options})")
```

Comment:
Yeah so this is my core question: does this extra-level of indirection here (query to df to query of DF that writes to an external file) introduce overhead or a perf hit that is going to cause issues for existing `external` users? Or is the DuckDB optimizer smart enough to optimize this away (for e.g. arrow record batches or some such thing?)

Reply:
Hi @jwills, a very good question
I am unsure if here goes and convert to arrow, but I would empirically try something today to showcase if this is the case. I intentionally hand over `DuckDBPyRelation ` so the plugin `store` function can choose the needed conversion.
Independent of the result, we would have a `target_config.config.model.compiled_code` variable, which we can embed into the string, and this should be the same as in the old path.

Reply:
Hi @jwills, in my unscientific empirical experiment, we can see there that the new flow/method is faster than the previous one, where we first create a table and then export it
I also added the conversion over the arrow and pandas data frame to make it better comparable 

![image](https://github.com/duckdb/dbt-duckdb/assets/46800472/44111b06-2364-4da7-b587-ca5339c2edda)

The code is here: https://github.com/milicevica23/wtf-memory/blob/main/002-basics-duckdb/export-parquet-benchmark.py

Please be aware that there can be some mistakes in terms of cache and swapping memory, but I think it is enough to show that there is no significant difference in the concept of how duckdb retrieves the data

Reply:
I also did a bit of experimenting with memory usage to see if i can visualize it; the b variant is better because there is no intermediate materialization 

version b)
![image](https://github.com/duckdb/dbt-duckdb/assets/46800472/e3569e9b-3060-4431-9609-2f206a1eddd1)

version c)
![Uploading image.png…]()